src/beanmachine/paic2/include/bm/bm_ops.td

```diff
@@ -0,0 +1,230 @@
+}
+
+// Base class for bm dialect operations. This operation inherits from the base
+// `Op` class in OpBase.td, and provides:
+//   * The parent dialect of the operation.
+//   * The mnemonic for the operation, or the name without the dialect prefix.
+//   * A list of traits for the operation.
+class BM_Op<string mnemonic, list<Trait> traits = []> :
+Op<BM_Dialect, mnemonic, traits>;
+
+// Provide a definition for the BM WorldType for use in ODS. This allows for
+// using WorldType in a similar way to Tensor or MemRef. We use `DialectType`
+// to demarcate the WorldType as belonging to the BM dialect.
+def BM_WorldType :
+DialectType<BM_Dialect, CPred<"$_self.isa<WorldType>()">,
+        "BM World type">;
+
+// this allows us to specify for example that a call operations has an out type of either BM_WorldType or F32.
+// in reality it can also return a pointer but we are not there yet.
+def BM_Type : AnyTypeOf<[F32, BM_WorldType]>;
+//===----------------------------------------------------------------------===//
+// WorldAccessOp
+//===----------------------------------------------------------------------===//
+
+def WorldAccessOp : BM_Op<"World_access", [NoSideEffect]> {
+    let summary = "World access";
+    let description = [{
+        Access the Nth element of a value returning a World type.
+    }];
+
+    let arguments = (ins BM_WorldType:$input, I64Attr:$index);
+    let results = (outs BM_Type:$output);
+
+    let assemblyFormat = [{
+        $input `[` $index `]` attr-dict `:` type($input) `->` type($output)
+    }];
+
+    // Allow building a WorldAccessOp with just a World value and an index.
+    let builders = [
+        OpBuilder<(ins "Value":$input, "size_t":$index)>
+    ];
+
+    // Set the folder bit so that we can fold constant accesses.
+    let hasFolder = 1;
+
+        // Indicate that additional verification for this operation is necessary.
+    let hasVerifier = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// WorldConstantOp
+//===----------------------------------------------------------------------===//
+
+def WorldConstantOp : BM_Op<"World_constant", [ConstantLike, NoSideEffect]> {
+    let summary = "World constant";
+    let description = [{
+        Constant operation turns a literal World value into an SSA value. The data
+        is attached to the operation as an attribute. The World constant is encoded
+        as an array of other constant values. For example:
+
+        ```mlir
+            %0 = bm.world_constant [
+            dense<[[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]> : tensor<2x3xf64>
+            ] : !bm.world<tensor<*xf64>>
+        ```
+    }];
+
+    let arguments = (ins ArrayAttr:$value);
+    let results = (outs BM_WorldType:$output);
+
+    let assemblyFormat = "$value attr-dict `:` type($output)";
+
+    // Indicate that additional verification for this operation is necessary.
+    let hasVerifier = 1;
+    let hasFolder = 1;
+}
```

Comment:
this op is also unused (although it will be used later, its not verified as working in this PR and should be removed)